docs/source/guide/events.rst

```diff
@@ -0,0 +1,445 @@
+
+    event_system.register('general', my_general_function)
+    event_system.register('general.specific', my_specific_function)
+
+    events_called = []
+    event_system.emit('general', my_list=events_called)
+    print(events_called)
+
+    events_called = []    
+    event_system.emit('general.specific', my_list=events_called)
+    print(events_called)
+
+Here is the output from running this example::
+
+    ['Called my general function']
+    ['Called my specific function', 'Called my general function']
+
+As expected, when the event ``'general'`` was emitted, only when the
+``my_general_function`` hander was called. However when the more specific event
+``'general.specific'`` was emitted, the ``my_specific_function`` and
+``my_general_function`` handlers were called in that order.
+
+
+Wildcard Matching
+-----------------
+
+Another aspect of Boto3's event system is that it has the capability
+to do wildcard matching using the ``'*'`` notation. Here is an example
+of using wildcards in the event system::
+
+    import boto3
+
+    s3 = boto3.client('s3')
+
+    # Access the event system on the S3 client
+    event_system = s3.meta.events
+
+    def my_general_function(my_list, **kwargs):
+        my_list.append('Called my general function')
+
+    def my_wildcard_function(my_list, **kwargs):
+        my_list.append('Called my wildcard function')
+
+    event_system.register('general', my_general_function)
+    event_system.register('general.*', my_wildcard_function)
+
+    events_called = []
+    event_system.emit('general', my_list=events_called)
+    print(events_called)
+
+    events_called = []    
+    event_system.emit('general.specific', my_list=events_called)
+    print(events_called)
+
+    events_called = []    
+    event_system.emit('general.other_specific', my_list=events_called)
+    print(events_called)
+
+Here is the output from running this example::
+
+    ['Called my general function']
+    ['Called my wildcard function', 'Called my general function']
+    ['Called my wildcard function', 'Called my general function']
+
+The ``'*'`` allows you to register to a group of events without having to
+know the actual name of the event. This is useful when you have to apply
+the same handler in multiple places. Also note that if the wildcard is used,
+it must be isolated. It does not handling globbing with additional characters.
+So in the previous example, if the ``my_wildcard_function`` was registered
+to ``'general.*specific'``, the handler would not be called because it will
+consider ``'general.*specific'`` to be a specific event.
+
+The wildcard also respects the hierarchical structure of the event system.
+So as shown in the previous example, the ``my_wildcard_function`` handler
+was called first because it is registered to a more specific event than the
+``my_function`` handler.
+
+
+Isolation of Event Emitter
+--------------------------
+
+The event system in Boto3 has the notion of isolation: registering of handlers
```

Comment:
This can be confusing when you take into account handlers registered on the session.  Maybe we can just simplify this by saying that all clients maintain their own set of registered handlers.